[CSS ZOOM] Apply zoom factor to border spacing and border width

nullhook opened the pull request:
#### 2ac6832444d1bbe2be30c4f0e0a0ab605c726058
<pre>
[CSS ZOOM] Apply zoom factor to border spacing and border width
<a href="https://bugs.webkit.org/show_bug.cgi?id=299238">https://bugs.webkit.org/show_bug.cgi?id=299238</a>
<a href="https://rdar.apple.com/161008191">rdar://161008191</a>

Reviewed by Antti Koivisto.

We allow evaluation-time zoom for LineWidth and WebKitBorderSpacing,
which applies the zoom factor to the border-width and border-spacing properties.

We also added a feature flag in this PR to enable evaluation-time zoom.

Test: imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width.html
* LayoutTests/TestExpectations:
* LayoutTests/platform/gtk/fast/multicol/span/anonymous-style-inheritance-expected.txt: Added.
* LayoutTests/platform/gtk/fast/repaint/repaint-during-scroll-with-zoom-expected.txt: Added.
* LayoutTests/platform/gtk/media/video-zoom-expected.txt: Added.
* LayoutTests/platform/gtk/svg/zoom/page/zoom-background-images-expected.txt: Added.
* LayoutTests/platform/gtk/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt: Added.
* LayoutTests/platform/gtk/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt: Added.
* LayoutTests/platform/gtk/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt: Added.
* LayoutTests/platform/ios/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/wpe/media/video-zoom-expected.txt: Added.
* LayoutTests/fast/backgrounds/size/contain-and-cover-zoomed-expected.txt:
* LayoutTests/fast/canvas/canvas-zoom-expected.html:
* LayoutTests/fast/sub-pixel/zoomed-em-border.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width.html: Added.
* LayoutTests/platform/ios/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/mac/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/mac/fast/multicol/span/anonymous-style-inheritance-expected.txt:
* LayoutTests/platform/mac/fast/repaint/repaint-during-scroll-with-zoom-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-background-images-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-float-border-padding-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt:
* LayoutTests/platform/wpe/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/wpe/fast/multicol/span/anonymous-style-inheritance-expected.txt: Added.
* LayoutTests/platform/wpe/fast/repaint/repaint-during-scroll-with-zoom-expected.txt: Copied from LayoutTests/platform/mac/fast/repaint/repaint-during-scroll-with-zoom-expected.txt.
* LayoutTests/platform/wpe/svg/zoom/page/zoom-background-images-expected.txt: Copied from LayoutTests/platform/mac/svg/zoom/page/zoom-background-images-expected.txt.
* LayoutTests/platform/wpe/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt: Copied from LayoutTests/platform/mac/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt.
* LayoutTests/platform/wpe/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt: Added.
* LayoutTests/platform/wpe/svg/zoom/page/zoom-svg-float-border-padding-expected.txt:
* LayoutTests/platform/wpe/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt: Copied from LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt.
* LayoutTests/svg/zoom/page/zoom-background-image-tiled-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/accessibility/AXTableHelpers.cpp:
(WebCore::AXTableHelpers::isDataTableWithTraversal):
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::computedBorder const):
* Source/WebCore/layout/formattingContexts/block/BlockFormattingGeometry.cpp:
(WebCore::Layout::BlockFormattingGeometry::intrinsicWidthConstraints const):
* Source/WebCore/layout/formattingContexts/table/TableFormattingState.cpp:
(WebCore::Layout::ensureTableGrid):
* Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp:
(WebCore::LayoutIntegration::BoxGeometryUpdater::logicalBorder):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fixedContainerEdges const):
* Source/WebCore/page/SpatialNavigation.cpp:
(WebCore::nodeRectInAbsoluteCoordinates):
* Source/WebCore/rendering/BorderEdge.cpp:
(WebCore::borderEdges):
(WebCore::borderEdgesForOutline):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintOutline const):
* Source/WebCore/rendering/BorderShape.cpp:
(WebCore::BorderShape::shapeForBorderRect):
* Source/WebCore/rendering/NinePieceImagePainter.cpp:
(WebCore::computeSlice):
(WebCore::paintNinePieceImage):
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
(WebCore::RenderBoxModelObject::borderAfter const):
(WebCore::RenderBoxModelObject::borderBefore const):
(WebCore::RenderBoxModelObject::borderBottom const):
(WebCore::RenderBoxModelObject::borderEnd const):
(WebCore::RenderBoxModelObject::borderLeft const):
(WebCore::RenderBoxModelObject::borderRight const):
(WebCore::RenderBoxModelObject::borderStart const):
(WebCore::RenderBoxModelObject::borderTop const):
(WebCore::RenderBoxModelObject::borderWidths const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::drawFocusRing):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintAreaElementFocusRing):
* Source/WebCore/rendering/RenderMultiColumnSet.cpp:
(WebCore::RenderMultiColumnSet::paintColumnRules):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::styleDidChange):
(WebCore::RenderTable::calcBorderStart const):
(WebCore::RenderTable::calcBorderEnd const):
(WebCore::RenderTable::outerBorderBefore const):
(WebCore::RenderTable::outerBorderAfter const):
(WebCore::RenderTable::outerBorderStart const):
(WebCore::RenderTable::outerBorderEnd const):
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::calcBlockDirectionOuterBorder const):
(WebCore::RenderTableSection::calcInlineDirectionOuterBorder const):
(WebCore::RenderTableSection::paintRowGroupBorderIfRequired):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::paintMenuListButtonDecorationsForVectorBasedControls):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::popupInternalPaddingBox const):
(WebCore::RenderThemeIOS::paintMenuListButtonDecorations):
* Source/WebCore/rendering/style/CollapsedBorderValue.h:
(WebCore::CollapsedBorderValue::CollapsedBorderValue):
(WebCore::CollapsedBorderValue::m_transparent): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setEnableEvaluationTimeZoom):
(WebCore::RenderStyle::imageOutsets const):
(WebCore::RenderStyle::outlineWidth const):
(WebCore::RenderStyle::outlineOffset const):
(WebCore::RenderStyle::outlineSize const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::enableEvaluationTimeZoom const):
(WebCore::RenderStyle::usedZoomForLength const):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeFloatVisibleRectInContainer):
* Source/WebCore/style/StyleResolveForDocument.cpp:
(WebCore::Style::resolveForDocument):
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.h:
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
(WebCore::Style::CSSValueConversion&lt;LineWidth&gt;::operator):
(WebCore::Style::FloatBoxExtent&gt;::operator):
(WebCore::Style::LayoutBoxExtent&gt;::operator):
* Source/WebCore/style/values/backgrounds/StyleLineWidth.h:
* Source/WebCore/style/values/masking/StyleMaskBorderWidth.h:
* Source/WebCore/style/values/non-standard/StyleWebKitBorderSpacing.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapperData.h:
(WebCore::Style::LengthWrapperData::minimumValueForLengthWrapperDataWithLazyMaximum const):
(WebCore::Style::LengthWrapperData::valueForLengthWrapperDataWithLazyMaximum const):
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StyleZoomNeededToken.h:
(WebCore::Style::ZoomFactor::ZoomFactor):
(WebCore::Style::ZoomFactor::operator float const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setEvaluationTimeZoom):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::absoluteInteractionBounds):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode innerFrameQuad]):

Canonical link: <a href="https://commits.webkit.org/300756@main">https://commits.webkit.org/300756@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5cd02b3971ce31adba416288e76cf48561d49ce9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123759 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43474 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34171 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/130537 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/75908 "Built successfully") | [❌ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d2c2a109-0343-4e91-a234-295e8c008b3b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125636 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44198 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52068 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94122 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/62461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c6fc40d6-fd8d-49e1-b08d-21516ec0ee3e) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126712 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35214 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/110717 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/74725 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/88a578be-e008-4091-9430-66fdf019b81c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34174 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/28876 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74017 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/115918 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/104938 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29100 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133227 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/122291 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/50711 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/38620 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/102598 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51086 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/106936 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102427 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26039 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/47769 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26022 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/47555 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/50565 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56329 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/153383 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50039 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/38992 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53385 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/51713 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/rendering/style/RenderStyle.cpp:
```diff
@@ -418,6 +418,11 @@ void RenderStyle::copyContentFrom(const RenderStyle& other)
     m_nonInheritedData.access().miscData.access().content = other.m_nonInheritedData->miscData->content;
 }
 
+void RenderStyle::setEnableEvaluationTimeZoom(bool value)
+{
+    SET_NESTED_VAR(m_nonInheritedData, rareData, enableEvaluationTimeZoom, value);
+}
```
I think you can make this an inherited rare data bit instead and then you'll only need to set it explicitly for the root style (in `Style::resolveForDocument()`).

nullhook commented on Source/WebCore/style/values/primitives/StyleZoomNeededToken.h:
```diff
@@ -30,5 +30,14 @@ namespace Style {
 // Token passed around to indicate that the evaluation will need zoom passed in the future.
 struct ZoomNeeded { };
 
+struct ZoomFactor {
```
added a dedicated type for zoom values instead of using floats.

when `R` has no zoom bit, `ZoomNeeded {}` is passed. if it does have a zoom bit, then `ZoomFactor` must be passed.

callers that have access RenderStyle object must use `usedZoomForLength` when zoom is required rather than usedZoom

weinig replied:
Either the file should be renamed, or this should go in its own file.

nullhook commented on Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
```diff
@@ -31,6 +31,8 @@
 #include "CalculationValue.h"
 #include "FloatConversion.h"
+#include "RenderStyleInlines.h"
+#include "Settings.h"
```
are adding these headers OK?

cc @weinig

nullhook replied:
actually, i can't add inlines header in this file.

nullhook replied:
fixed.

anttijk commented on Source/WebCore/style/StyleResolveForDocument.cpp:
```diff
@@ -108,6 +108,9 @@ RenderStyle resolveForDocument(const Document& document)
     documentStyle.setFontCascade(WTFMove(fontCascade));
 
+    if (document.settings().evaluationTimeZoomEnabled())
+        documentStyle.setEnableEvaluationTimeZoom(true);
```
could just say
 ```
documentStyle.setEnableEvaluationTimeZoom(document.settings().evaluationTimeZoomEnabled());
```

nullhook committed:
[CSS ZOOM] Apply zoom factor to border spacing and border width
https://bugs.webkit.org/show_bug.cgi?id=299238
rdar://161008191

Reviewed by Antti Koivisto.

We allow evaluation-time zoom for LineWidth and WebKitBorderSpacing,
which applies the zoom factor to the border-width and border-spacing properties.

We also added a feature flag in this PR to enable evaluation-time zoom.

Test: imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width.html
* LayoutTests/TestExpectations:
* LayoutTests/platform/gtk/fast/multicol/span/anonymous-style-inheritance-expected.txt: Added.
* LayoutTests/platform/gtk/fast/repaint/repaint-during-scroll-with-zoom-expected.txt: Added.
* LayoutTests/platform/gtk/media/video-zoom-expected.txt: Added.
* LayoutTests/platform/gtk/svg/zoom/page/zoom-background-images-expected.txt: Added.
* LayoutTests/platform/gtk/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt: Added.
* LayoutTests/platform/gtk/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt: Added.
* LayoutTests/platform/gtk/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt: Added.
* LayoutTests/platform/ios/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/wpe/media/video-zoom-expected.txt: Added.
* LayoutTests/fast/backgrounds/size/contain-and-cover-zoomed-expected.txt:
* LayoutTests/fast/canvas/canvas-zoom-expected.html:
* LayoutTests/fast/sub-pixel/zoomed-em-border.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-viewport/zoom/border-width.html: Added.
* LayoutTests/platform/ios/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/mac/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/mac/fast/multicol/span/anonymous-style-inheritance-expected.txt:
* LayoutTests/platform/mac/fast/repaint/repaint-during-scroll-with-zoom-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-background-images-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-float-border-padding-expected.txt:
* LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt:
* LayoutTests/platform/wpe/fast/backgrounds/background-position-parsing-expected.txt:
* LayoutTests/platform/wpe/fast/multicol/span/anonymous-style-inheritance-expected.txt: Added.
* LayoutTests/platform/wpe/fast/repaint/repaint-during-scroll-with-zoom-expected.txt: Copied from LayoutTests/platform/mac/fast/repaint/repaint-during-scroll-with-zoom-expected.txt.
* LayoutTests/platform/wpe/svg/zoom/page/zoom-background-images-expected.txt: Copied from LayoutTests/platform/mac/svg/zoom/page/zoom-background-images-expected.txt.
* LayoutTests/platform/wpe/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt: Copied from LayoutTests/platform/mac/svg/zoom/page/zoom-img-preserveAspectRatio-support-1-expected.txt.
* LayoutTests/platform/wpe/svg/zoom/page/zoom-replaced-intrinsic-ratio-001-expected.txt: Added.
* LayoutTests/platform/wpe/svg/zoom/page/zoom-svg-float-border-padding-expected.txt:
* LayoutTests/platform/wpe/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt: Copied from LayoutTests/platform/mac/svg/zoom/page/zoom-svg-through-object-with-auto-size-expected.txt.
* LayoutTests/svg/zoom/page/zoom-background-image-tiled-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/accessibility/AXTableHelpers.cpp:
(WebCore::AXTableHelpers::isDataTableWithTraversal):
* Source/WebCore/layout/formattingContexts/FormattingGeometry.cpp:
(WebCore::Layout::FormattingGeometry::computedBorder const):
* Source/WebCore/layout/formattingContexts/block/BlockFormattingGeometry.cpp:
(WebCore::Layout::BlockFormattingGeometry::intrinsicWidthConstraints const):
* Source/WebCore/layout/formattingContexts/table/TableFormattingState.cpp:
(WebCore::Layout::ensureTableGrid):
* Source/WebCore/layout/integration/LayoutIntegrationBoxGeometryUpdater.cpp:
(WebCore::LayoutIntegration::BoxGeometryUpdater::logicalBorder):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::fixedContainerEdges const):
* Source/WebCore/page/SpatialNavigation.cpp:
(WebCore::nodeRectInAbsoluteCoordinates):
* Source/WebCore/rendering/BorderEdge.cpp:
(WebCore::borderEdges):
(WebCore::borderEdgesForOutline):
* Source/WebCore/rendering/BorderPainter.cpp:
(WebCore::BorderPainter::paintOutline const):
* Source/WebCore/rendering/BorderShape.cpp:
(WebCore::BorderShape::shapeForBorderRect):
* Source/WebCore/rendering/NinePieceImagePainter.cpp:
(WebCore::computeSlice):
(WebCore::paintNinePieceImage):
* Source/WebCore/rendering/RenderBoxModelObjectInlines.h:
(WebCore::RenderBoxModelObject::borderAfter const):
(WebCore::RenderBoxModelObject::borderBefore const):
(WebCore::RenderBoxModelObject::borderBottom const):
(WebCore::RenderBoxModelObject::borderEnd const):
(WebCore::RenderBoxModelObject::borderLeft const):
(WebCore::RenderBoxModelObject::borderRight const):
(WebCore::RenderBoxModelObject::borderStart const):
(WebCore::RenderBoxModelObject::borderTop const):
(WebCore::RenderBoxModelObject::borderWidths const):
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::drawFocusRing):
* Source/WebCore/rendering/RenderImage.cpp:
(WebCore::RenderImage::paintAreaElementFocusRing):
* Source/WebCore/rendering/RenderMultiColumnSet.cpp:
(WebCore::RenderMultiColumnSet::paintColumnRules):
* Source/WebCore/rendering/RenderTable.cpp:
(WebCore::RenderTable::styleDidChange):
(WebCore::RenderTable::calcBorderStart const):
(WebCore::RenderTable::calcBorderEnd const):
(WebCore::RenderTable::outerBorderBefore const):
(WebCore::RenderTable::outerBorderAfter const):
(WebCore::RenderTable::outerBorderStart const):
(WebCore::RenderTable::outerBorderEnd const):
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::calcBlockDirectionOuterBorder const):
(WebCore::RenderTableSection::calcInlineDirectionOuterBorder const):
(WebCore::RenderTableSection::paintRowGroupBorderIfRequired):
* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::extractControlStyleForRenderer const):
(WebCore::RenderTheme::adjustButtonOrCheckboxOrColorWellOrInnerSpinButtonOrRadioStyle const):
* Source/WebCore/rendering/cocoa/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::paintMenuListButtonDecorationsForVectorBasedControls):
* Source/WebCore/rendering/ios/RenderThemeIOS.mm:
(WebCore::RenderThemeIOS::popupInternalPaddingBox const):
(WebCore::RenderThemeIOS::paintMenuListButtonDecorations):
* Source/WebCore/rendering/style/CollapsedBorderValue.h:
(WebCore::CollapsedBorderValue::CollapsedBorderValue):
(WebCore::CollapsedBorderValue::m_transparent): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::setEnableEvaluationTimeZoom):
(WebCore::RenderStyle::imageOutsets const):
(WebCore::RenderStyle::outlineWidth const):
(WebCore::RenderStyle::outlineOffset const):
(WebCore::RenderStyle::outlineSize const):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
(WebCore::RenderStyle::enableEvaluationTimeZoom const):
(WebCore::RenderStyle::usedZoomForLength const):
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
(WebCore::StyleRareInheritedData::StyleRareInheritedData):
(WebCore::StyleRareInheritedData::operator== const):
(WebCore::StyleRareInheritedData::dumpDifferences const):
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::computeFloatVisibleRectInContainer):
* Source/WebCore/style/StyleResolveForDocument.cpp:
(WebCore::Style::resolveForDocument):
* Source/WebCore/style/values/backgrounds/StyleBorderImageWidth.h:
* Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
(WebCore::Style::CSSValueConversion<LineWidth>::operator):
(WebCore::Style::FloatBoxExtent>::operator):
(WebCore::Style::LayoutBoxExtent>::operator):
* Source/WebCore/style/values/backgrounds/StyleLineWidth.h:
* Source/WebCore/style/values/masking/StyleMaskBorderWidth.h:
* Source/WebCore/style/values/non-standard/StyleWebKitBorderSpacing.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapper.h:
* Source/WebCore/style/values/primitives/StyleLengthWrapperData.h:
(WebCore::Style::LengthWrapperData::minimumValueForLengthWrapperDataWithLazyMaximum const):
(WebCore::Style::LengthWrapperData::valueForLengthWrapperDataWithLazyMaximum const):
* Source/WebCore/style/values/primitives/StylePrimitiveNumeric.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+CSSValueConversion.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Conversions.h:
* Source/WebCore/style/values/primitives/StylePrimitiveNumericTypes+Evaluation.h:
* Source/WebCore/style/values/primitives/StyleZoomNeededToken.h:
(WebCore::Style::ZoomFactor::ZoomFactor):
(WebCore::Style::ZoomFactor::operator float const):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::setEvaluationTimeZoom):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::absoluteInteractionBounds):
* Source/WebKitLegacy/mac/DOM/DOM.mm:
(-[DOMNode innerFrameQuad]):

Canonical link: https://commits.webkit.org/300756@main

weinig commented on Source/WebCore/rendering/style/CollapsedBorderValue.h:
```diff
@@ -39,7 +39,7 @@ class CollapsedBorderValue {
 
     CollapsedBorderValue(const BorderValue& border, const Color& color, BorderPrecedence precedence)
-        : m_width(border.nonZero() ? Style::evaluate<LayoutUnit>(border.width(), Style::ZoomNeeded { }) : 0_lu)
+        : m_width(border.nonZero() ? Style::evaluate<LayoutUnit>(border.width(), Style::ZoomFactor(1.0f)) : 0_lu)
```
I don't understand this? Why is 1.0f being passed here.

nullhook replied:
that was supposed to be addressed later, but the change resulted in a long PR, so it was missed. apologies.

perhaps this should take ZoomFactor?

```
CollapsedBorderValue(const BorderValue& border, const Color& color, BorderPrecedence precedence, const Style::ZoomFactor& zoom)
```

can also take a Style object but i see call-sites like the following that requires a unity factor.


```
static CollapsedBorderValue emptyBorder()
{
    return CollapsedBorderValue(BorderValue(), Color(), BorderPrecedence::Cell, Style::ZoomFactor(1.0f));
}
```

weinig replied:
I think it would be bad to find ourselves passing in Style::ZoomFactor(1.0f) to places, so let's try to avoid that.

In this case, the constructor immediately evaluates it, so one approach could be to just stop passing in a BorderValue and instead pass in the resolved bits.

```
CollapsedBorderValue(const LayoutUnit& borderWidth, BorderStyle borderStyle, bool isTransparent, const Color& color, BorderPrecedence precedence).
```

nullhook replied:
i'm not sure i understand the reasoning behind avoiding `Style::ZoomFactor(1.0f)` in this specific case. the example i referenced is emptyBorder(), which is a static function that would use a zoom factor of 1.0f regardless, since it has no access to any element's style context.

also, your suggestion to change the constructor parameters and pass "resolved bits" instead would be significantly more involved, For example, looking at `RenderTableCell::computeCollapsedStartBorder()`, there are many cases that would need to be updated, and this approach would be very easy to get wrong.

nullhook replied:
for ex. the following derives border definition from row's style and color from parent's style. 

```
result = chooseBorder(result, CollapsedBorderValue(row()->borderAdjoiningEndCell(*this), includeColor ?
  parent()->style().visitedDependentColorWithColorFilter(endColorProperty) : Color(), BorderPrecedence::Row));
```

nullhook replied:
i guess the answer above is the border and zoom should be derived from same style object

weinig commented on Source/WebCore/testing/Internals.h:
```diff
@@ -1629,6 +1629,7 @@ class Internals final
     bool isModelElementIntersectingViewport(HTMLModelElement&);
 #endif
 
+    void setEvaluationTimeZoom(bool value);
```
This should not be needed. Settings are all exposed automatically.

weinig commented on Source/WebCore/rendering/NinePieceImagePainter.cpp:
```diff
@@ -79,7 +79,7 @@ static LayoutUnit computeSlice(const WidthValue& length, LayoutUnit width, Layou
     return WTF::switchOn(length,
         [&](const typename WidthValue::LengthPercentage& value) {
-            return Style::evaluate<LayoutUnit>(value, extent, Style::ZoomNeeded { });
+            return Style::evaluate<LayoutUnit>(value, extent, Style::ZoomFactor(1.0f) /* FIXME FIND ZOOM */);
```
The whole point of the `Style::ZoomNeeded { }` is to get rid of the /* FIXME FIND ZOOM */. Why is this being added back?

nullhook replied:
fixed.

weinig commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1287,7 +1287,7 @@ LayoutUnit RenderTable::calcBorderStart() const
         }
     }
-    return CollapsedBorderValue::adjustedCollapsedBorderWidth(Style::evaluate<float>(borderWidth, Style::ZoomNeeded { }), document().deviceScaleFactor(), writingMode().isInlineFlipped());
+    return CollapsedBorderValue::adjustedCollapsedBorderWidth(Style::evaluate<float>(borderWidth, style().usedZoomForLength()), document().deviceScaleFactor(), writingMode().isInlineFlipped());
```
I don't believe this (or many of the others in this file) are using the correct zoom here. 

This `borderWidth` may come from one of multiple different RenderObjects (this, column, topNonEmptySection, and adjoiningStartCell), each with their own RenderStyle and zoom values.

The correct one must be paired here.

nullhook replied:
good spot.

i think outerBorderBefore, outerBorderStart are correct as they use `tb` which is table border and i'm reading `style()` here. 

however, i see for `calcBorderStart`, and `calcBorderEnd` evaluation is problematic as there are many sources of border widths.

nullhook replied:
fixed.

weinig commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -1364,4 +1364,17 @@ inline bool RenderStyle::fontCascadeEqual(const RenderStyle& other) const
+
+inline Style::ZoomFactor RenderStyle::usedZoomForLength() const
+{
+    if (enableEvaluationTimeZoom())
```
This doesn't seem to be taking into account SVG zoom rules.

nullhook replied:
tagging RenderStyle object with useSVGZoomRulesForLength bit here.

weinig commented on Source/WebCore/rendering/ios/RenderThemeIOS.mm:
```diff
@@ -610,9 +610,9 @@ static void adjustInputElementButtonStyle(RenderStyle& style, const HTMLInputEle
     glyphOrigin.setY(logicalRect.center().y() - glyphSize.height() / 2.0f);
     if (!style.writingMode().isInlineFlipped())
-        glyphOrigin.setX(logicalRect.maxX() - glyphSize.width() - Style::evaluate<float>(box.style().borderEndWidth(), Style::ZoomNeeded { }) - Style::evaluate<float>(box.style().paddingEnd(), logicalRect.width(), Style::ZoomNeeded { }));
+        glyphOrigin.setX(logicalRect.maxX() - glyphSize.width() - Style::evaluate<float>(box.style().borderEndWidth(), style.usedZoomForLength()) - Style::evaluate<float>(box.style().paddingEnd(), logicalRect.width(), Style::ZoomNeeded { }));
```
This is using a mix of `box.style()` and `style`. Are they the same? If so, why two different ways to reference them?

nullhook replied:
sorry, i didn't link the PR that actually solves this: https://github.com/WebKit/WebKit/pull/51583

weinig commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -30,6 +30,7 @@
 #include <WebCore/PseudoElementIdentifier.h>
 #include <WebCore/StylePrimitiveNumeric+Forward.h>
 #include <WebCore/StyleTextDecorationLine.h>
+#include <WebCore/StyleWebKitBorderSpacing.h>
```
Please don't add new includes to RenderStyle.h. Forward declarations should work fine.

weinig commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -408,7 +409,6 @@ using TransformOriginXY = Position;
 using TransformOriginY = PositionY;
 using TransformOriginZ = Length<>;
 using Transitions = CoordinatedValueList<Transition>;
-using WebkitBorderSpacing = Length<CSS::Nonnegative>;
```
Let's keep this forward declared. If you need to change the range in two places, that's ok.

weinig commented on Source/WebCore/rendering/style/StyleRareInheritedData.h:
```diff
@@ -230,7 +230,7 @@ class StyleRareInheritedData : public RefCounted<StyleRareInheritedData> {
     Style::ListStyleType listStyleType;
     Style::BlockEllipsis blockEllipsis;
-
+    PREFERRED_TYPE(bool) unsigned enableEvaluationTimeZoom : 1;
```
Why is this inherited data?

nullhook replied:
because of the following feedback: https://github.com/WebKit/WebKit/pull/51076#discussion_r2388684139

weinig commented on Source/WebCore/style/values/backgrounds/StyleLineWidth.cpp:
```diff
@@ -57,40 +57,38 @@ auto CSSValueConversion<LineWidth>::operator()(BuilderState& state, const CSSVal
     }
 
-    // Any original result that was >= 1 should not be allowed to fall below 1. This keeps border lines from vanishing.
+    // For LineWidth, we need special logic to prevent lines from vanishing
```
Should the settings value be checked here?

weinig commented on Source/WebCore/style/values/primitives/StyleZoomNeededToken.h:
```diff
@@ -30,5 +30,14 @@ namespace Style {
+struct ZoomFactor {
+    float value;
+
+    constexpr ZoomFactor() : value(1.0f) { }
```
I don't think default construction is wise here. It would almost always be an error to use this, so we might as well make callers do it explicitly by passing 1.0f if they really want it.

weinig commented on Source/WebCore/style/values/primitives/StyleZoomNeededToken.h:
```diff
@@ -30,5 +30,14 @@ namespace Style {
+    constexpr ZoomFactor() : value(1.0f) { }
+    constexpr explicit ZoomFactor(float v) : value(v) { }
+
+    constexpr operator float() const { return value; }
```
I think you are better off without this for now. Not that many places need to evaluate the actual zoom value, and this will create the possibility of more overload ambiguity.

nullhook replied:
what is 'this' referring to?

weinig replied:
The line of code. The "operator float".
